bitcoin/src/bip32.rs

```diff
@@ -111,55 +113,145 @@ pub struct Xpub {
 #[cfg(feature = "serde")]
 internals::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 
-/// A child number for a derived key
+/// A BIP32 types of key derivation
 #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
 pub enum ChildNumber {
-    /// Non-hardened key
-    Normal {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
-    /// Hardened key
-    Hardened {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
+    /// Enum wrapper for a normal child number
+    Normal(NormalChildNumber),
+    /// Enum wrapper for a hardened child number
+    Hardened(HardenedChildNumber),
+}
+///Sub-module for HardenedChildNumber type
+pub mod hardened {
+    use super::*;
+    /// A child number for a derived key
+    /// Indexes 2147483648 to 4294967295 (the second half of all possible children)
+    #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+    pub struct HardenedChildNumber(u32);
+
+    impl HardenedChildNumber {
+        /// `HardenedChildNumber` with index 0.
```

Comment:
I having intrusive thoughts to make this impl a trait for numbers... but since custom derivation key derivation are not a good pratice while developing a wallet sounds like an useless change

Reply:
You could make an argument that increasing the index by one is a standard operation that wallets do but to implement `Add` just for that looks too much. Also what argument would you use? Neither `u32` nor `Self` makes sense to me.

Reply:
Yep, doesnt for me too.